### ISSUE TITLE
Track gaps on Android

### ISSUE THREAD

**Issue by losnikitos** [transistorsoft/react-native-background-geolocation#699]
## Your Environment
* Plugin version: 3.0.0-rc.4
* Platform: Android
* OS version: 9.0
* Device manufacturer / model: Samsung Galaxy S9 
* React Native version (`react-native -v`): 0.55.4
* Plugin config
```js
{
    desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
    distanceFilter: 10,
    fastestLocationUpdateInterval: 1000,
    locationUpdateInterval: 10000,
    heartbeatInterval: 60,
    stopOnTerminate: false,
    startOnBoot: true,
    foregroundService: true,
    notificationTitle: 'Yandex.Courier',
    notificationText: 'Uses Geolocation',
    maxDaysToPersist: 3,
    url: '...',
    autoSync: true,
    batchSync: true,
    maxBatchSize: 100,
    httpRootProperty: 'positions',
    locationTemplate: '{ "accuracy": <%= accuracy %>, "latitude":<%= latitude %>, "longitude":<%= longitude %>, "time": "<%= timestamp %>" }',
    headers: {
      Authorization: `Auth ${token}`,
      'User-Agent': `com.yandex.courier v${pkg.version}`,
    }
  };
```
## Expected Behavior
Location updates recorded every 10 meters, locations pushed to http server

## Actual Behavior
Location updates are not recorded nor pushed to server until phone is unlocked and app is in foreground

## Steps to Reproduce

1.
2.
3.
4.

## Context
We're developing app for couriers with tracking support. Sometimes we see gaps in tracks up to several hours. In logs there's nothing special, no errors. Users use different devices, we've reproduced the issue on one device.

Our best guess is that OS kills background geolocation process occasionally.

## Debug logs
We've recorded a track with 5 gaps, each around 8 minutes long.

#### Locations log
https://gist.github.com/losnikitos/dfe6fc6a1d03f997c50cc74848c3b65f

#### Track
![](https://yadi.sk/i/z37M8QTxV1D6yQ_xxl.jpg)

#### logcats:
14:42:13 - 14:51:04 (9 minutes) https://gist.github.com/losnikitos/df064563a4681c2ffac57bde5f6e960f
14:51:16 - 14:59:45 (8 minutes) https://gist.github.com/losnikitos/73868dbb4fc937c3f04f7bb923af7fad
14:59:59 - 15:08:29 (8 minutes) https://gist.github.com/losnikitos/e83c57d5a4fd08e3ba876cf6513c7f7b
15:08:29 - 15:16:52 (8 minutes) https://gist.github.com/losnikitos/806b4dd012e6782cfc4ab8905cd0d9a6
15:17:04 - 15:25:20 (8 minutes) https://gist.github.com/losnikitos/1876c6b2412c5ee749b2faee7290a008

**Comment by christocracy** [comment]
It's impossible to see individual locations on your screenshot.

Post to my demo server.

See docs [transistorTrackerParams](https://transistorsoft.github.io/react-native-background-geolocation/classes/_react_native_background_geolocation_.backgroundgeolocation.html#transistortrackerparams)

**Comment by christocracy** [comment]
and set your `locationUpdateInterval: 1000` instead of `10000`.

**Comment by losnikitos** [comment]
@christocracy Thank you, we'll post new track to your server after the weekend

**Comment by UnSket** [comment]
We got response from test server with code status 500. Should we set some else configs like locationTemplate?

**Comment by christocracy** [comment]
You can't use `locationTemplate` with the test server.  How could it possibly be aware of your specific custom template and where to find `latitude`, `longitude`, etc?

Set `locationTemplate: ''` when posting to test-sever.

**Comment by UnSket** [comment]
I can`t send data to your test server. Always got 500 with 'Something failed!' label. Below some information with details:

- I use only two params like in docs : 
`{
    url: 'http://tracker.transistorsoft.com/locations/unsket',
    params: BackgroundGeolocation.transistorTrackerParams(DeviceInfo)
  }`
- Plugin version: '3.0.0-rc.4'
- Result of BackgroundGeolocation.transistorTrackerParams(device):
`{ uuid: 'ONEPLUS-A6003',
         model: 'ONEPLUS A6003',
         platform: 'Android',
         manufacturer: 'OnePlus',
         version: '8.1.0',
         framework: 'ReactNative' } }`

Maybe you know what problem could it be @christocracy?

**Comment by christocracy** [comment]
When you *change* config supplied to `#ready`, it's loaded only on first-launch of your app.  While developing, provide `reset: true`.

**Comment by UnSket** [comment]
I reinstall my app with that flag. It's work. Thanks!

**Comment by christocracy** [comment]
You didn't need to re-install, just set `reset: true` and re-launch the app.

**Comment by UnSket** [comment]
I did it at first, but it worked only after reinstalling

**Comment by christocracy** [comment]
You did something wrong.  It's perfectly fine to *always* use `reset: true` so that `#ready` always applies the supplied config.

**Comment by UnSket** [comment]
I'll check it. Anyway now it works and we will send you track soon. Thanks

**Comment by UnSket** [comment]
http://tracker.transistorsoft.com/unsket1?device=ReactNative-SM-G965F&end=2019-5-6&start=2019-5-6%200%3A1 - here is the link to the problem track. I also recorded it in Google Maps. Below is the track from it.
![google-track](https://user-images.githubusercontent.com/31548789/57227447-1a0c8700-701a-11e9-83c5-5eed400af086.jpg)

**Comment by christocracy** [comment]
Look at your `accuracy`:  your recorded locations are *not* coming from GPS.  When locations come from GPS, you'll see `accuracy <= 10` and the `heading` / `speed` will not be `-1`.

![](https://dl.dropbox.com/s/d1m7fohwc4atsyp/Screenshot%202019-05-06%2009.27.06.png?dl=1)

You have either:
- disabled GPS on that device, via *Settings->Location->Mode->Battery Saving* (should be set to *High Accuracy*) 
- Configured the plugin with `desiredAccuracy` not set to `DESIRED_ACCURACY_HIGH` or `DESIRED_ACCURACY_NAVIGATION`.

**Comment by UnSket** [comment]
If i disable GPS there will not any locations on map (i tried it). I checked seetting with accuracy it is high (desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH).

Could be something else?

**Comment by christocracy** [comment]
>Could be something else?

Yes.  The environment (eg: tall buildings, mountains, large bodies of water).  Broken device.  Verify your `state.desiredAccuracy`

```javascript
BackgroundGeolocation.ready(config).then((state) => {
  // Should be < 10 (eg: -2)
  console.log('[ready] desiredAccuracy: ', state.desiredAccuracy);
});

**Comment by UnSket** [comment]
> Yes. The environment (eg: tall buildings, mountains, large bodies of water). Broken device. Verify your `state.desiredAccuracy`
But second phone have good track. And I'm pretty sure that if I ran the application, the track would start to draw normally. 
We have same problem on many devices. Look like it is not broken phone

Result of console log:
`'[ready] desiredAccuracy: ', -1`

**Comment by UnSket** [comment]
I also walked with launched app with that phone:
http://tracker.transistorsoft.com/unsket?device=ReactNative-ONEPLUS-A6003&end=2019-5-6%2017%3A0&start=2019-5-6%2016%3A45

**Comment by christocracy** [comment]
The plugin's `desiredAccuracy` is correctly configured for `HIGH_ACCURACY`.

The problem is the device.
https://dontkillmyapp.com/

My [OnePlus A5010](http://tracker.transistorsoft.com/transistor-flutter?device=Flutter-ONEPLUS-A5010-9&end=2019-5-5&start=2019-4-25) consistently performs perfectly.

**Comment by UnSket** [comment]
I also went with the application running in one direction and closed in the opposite direction: 
http://tracker.transistorsoft.com/unsket1?device=ReactNative-SM-G965F&end=2019-5-6%2017%3A20&start=2019-5-6%2016%3A45

**Comment by UnSket** [comment]
I have gaps like this on other devices for example at ONEPLUS A6003. I tried to do all from advices from dontkillmyapp. It is not helped.

**Comment by christocracy** [comment]
I don't know what else to tell you.

**Comment by UnSket** [comment]
If we look at logcat logs we will see next line before gap:
`04-22 14:51:17.400 30113 30321 I TSLocationManager: [c.t.l.u.BackgroundTaskManager$Task stop] ⏳ stopBackgroundTask: 2`
And start service time after it is unpredictable. 
Could it be unreliable background task scheduling?

**Comment by christocracy** [comment]
⏳ stopBackgroundTask

You’ll see one of these with each http request.

Drawing conclusions on that is irrelevant.

**Comment by losnikitos** [comment]
Hi! We have recorded a few more tracks with gaps:

1. OnePlus 5T, Android 9.0.5
	http://tracker.transistorsoft.com/lamo?device=ReactNative-ONEPLUS-A5010&end=2019-5-13&start=2019-5-13%208%3A0

	Actual route https://yandex.ru/maps/213/moscow/?ll=37.446378%2C55.844168&mode=routes&rtext=55.843874%2C37.445347~55.771052%2C37.593821~55.734739%2C37.588449&rtt=auto&z=16.94

	* Gap between first two points 08:41 to 08:46 UTC.
	* Gap between 09:10 and 09:13 UTC (track stops when user is inside tunnel, but doesn't continue when he is outside again).
	* No locations recoded between May, 10 and May, 12. The phone was on, app was launched and track was recoded since May, 9. But only part of track on May, 9 and May, 13 were recorded.

2. OnePlus 6, Android 8.1.0
	http://tracker.transistorsoft.com/unsket?device=ReactNative-ONEPLUS-A6003&end=2019-5-11%2015%3A0&start=2019-5-11

	* Gap between 10:09 and 10:25 UTC (points 98784ac7-27cf-45b0-9f1e-27ec7b9d7855 and a6039410-9f64-4e27-96bd-38d94d522c73)

3. Samsung 9S+, Android 9.0.0
	http://tracker.transistorsoft.com/unsket1?device=ReactNative-SM-G965F&end=2019-5-12%200%3A30&start=2019-5-11%2023%3A40

	* Track is not continuous. Compare with same track recoded on OnePlus: http://tracker.transistorsoft.com/unsket?device=ReactNative-ONEPLUS-A6003&end=2019-5-12%200%3A30&start=2019-5-11%2023%

4. Samsung 9S+, Android 9.0.0
	http://tracker.transistorsoft.com/unsket1?device=ReactNative-SM-G965F&end=2019-5-6&start=2019-5-6%200%3A1

	* Track is not continuous. Compare with same track from OnePlus: http://tracker.transistorsoft.com/unsket?device=ReactNative-ONEPLUS-A6003&end=2019-5-6&start=2019-5-6%2013%3A0

-------

In previous comments you have advised us to check plugin settings and environment. We checked that:

> disabled GPS on that device, via Settings->Location->Mode->Battery Saving (should be set to High Accuracy)

We have `Mode: High Accuracy` on our test devices, where this setting is present

> Configured the plugin with desiredAccuracy not set to DESIRED_ACCURACY_HIGH or DESIRED_ACCURACY_NAVIGATION.

We have `DESIRED_ACCURACY_HIGH`

> The environment (eg: tall buildings, mountains, large bodies of water)

We've tried to record tracks in different conditions. Track 4 is recoded in open air, in a park https://goo.gl/maps/bpnS5QTARFQoKjmh9 and it still has gaps.

> Broken device

We've tried 3 devices ourselves, we have many reports from users. I don't believe all these devices are broken.

> Verify your `state.desiredAccuracy`, should be < 10

We have `-1`.

> The problem is the device. https://dontkillmyapp.com/

We've followed the instructions there (turned off battery saving for our app). We've tried both 'bad' (OnePlus) and 'good' (Samsung) devices from the list. Our users have various devices, they have track gaps on Philips S318 (Android 7.0), HUAWEI Y5 2018 (Android 8.1.0) and Galaxy A5(2017) (Android 8.0.0).

------

Looks like none of the advice above helps. We still have gaps in tracks.

We make courier delivery tracking system. We mark orders delivered based on courier's position, so it's important for us that track is continuous. 

@christocracy, is there something else we can do? We can provide any logs, we have QA engineers team, they have many devices, they can test the app under certain conditions. Since we don't have the source code, we cannot dig deeper into the problem, we only can observe tracks and logs, and they don't give any clue.

**Comment by UnSket** [comment]
We also found an interesting relationship: top of our phones with bad track has A-GPS. Could there be a problem?

**Comment by christocracy** [comment]
>Could there be a problem?

Could be.  I'm sure the problem is device-related.  There's nothing the plugin can do.

**Comment by r3dm1ke** [comment]
Sorry for intruding, but I noticed you are trying to track GPS in the Moscow, and there are a lot of issues with it and that could be a factor (https://insfinance.ru/36995-v-moskve-opyat-slomalsya-gps-na-etot-raz-po-vsemu-gorodu.html , https://www.gazeta.ru/auto/2016/12/16_a_10430909.shtml). Just try using any other tracking application and see if this behavior is consistent.

**Comment by christocracy** [comment]
@UnSket And be sure to use the latest [release](https://github.com/transistorsoft/cordova-background-geolocation/releases) of the plugin.

**Comment by losnikitos** [comment]
@r3dm1ke you're welcome to intrude :) Yes, I know about poor GPS quality and jumps to airports near Kremlin, but we have problems not just in the center of Moscow. We have bugreports from Tver, Izhevsk. We've recorded tracks from two devices at the same time; if satellite signal was bad, both would have bad tracks, but we don't see it. There's also this this 2-days gap when our teammate was travelling between cities, and most of the track was not recorded — it's not just a problem in Moscow.

@christocracy 
> I'm sure the problem is device-related. There's nothing the plugin can do.
Can the plugin write warnings to logs, when background process was killed? We've tried time-based tracking, setting `locationUpdateInterval` to 10 seconds; if location was not received after e.g. 60 seconds, maybe there should be a warning? Heartbeat events also stop at some point. Maybe there must be something in logs about it?

My point is that device might behave wrong, but at least I want to know from the logs, that something went wrong. Then when we see similar reports, we might see a pattern. Right now I don't see any pattern: we have different cities, different device manufacturers and OS versions.

**Comment by UnSket** [comment]
@christocracy Hi! We found new issue with plugin:
![image](https://user-images.githubusercontent.com/31548789/59502587-c6a61800-8ea6-11e9-898e-8a9f227413cd.png)
Is it ok? Looks like gaps could be created by this.

**Comment by christocracy** [comment]
@UnSket what version of the plugin are those errors coming from?  It's should be impossible to generate a `NullPointerException` there.

```java
class BackgroundTaskManager {
    .
    .
    .
    /**
     * Signal from BackgroundTaskService that the BackgroundTaskService is booted and ready to execute tasks.
     * @param context
     * @param taskId
     * @param completionHandler
     */
    public void onStartJob(Context context, int taskId, Callback completionHandler) {
        Task task = find(taskId);
        if (task == null) {  // <-- if task is null, bail out.
            TSLog.logger.warn(TSLog.warn("Failed to find Task: " + taskId));
            completionHandler.onFinish();
            return;
        }
        TSLog.logger.info(TSLog.ICON_HOURGLASS + " startBackgroundTask: " + taskId);

        // Task is guaranteed NOT NULL here.  This is where you NullPointerException is generated.
        task.start(completionHandler);
    }
}
```

**Comment by UnSket** [comment]
It is 3.0.0-rc.4 now. Should we update to another version? It started in April then was version 2+. Interesting fact that is most affected devices is HUAWEI.
![image](https://user-images.githubusercontent.com/31548789/59551127-7c439a80-8f7d-11e9-897c-e20e273de67d.png)

**Comment by UnSket** [comment]
Here is full stacktrace:Caused by java.lang.NullPointerException: Attempt to invoke virtual method 'void com.transistorsoft.locationmanager.util.BackgroundTaskManager$Task.start(android.content.Context, com.transistorsoft.locationmanager.util.BackgroundTaskManager$a)' on a null object reference
       at com.transistorsoft.locationmanager.util.BackgroundTaskManager.onStartJob(Unknown Source:30)
       at com.transistorsoft.locationmanager.service.BackgroundTaskService.onStartCommand(Unknown Source:40)
       at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4222)
       at android.app.ActivityThread.access$2500(ActivityThread.java:268)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2026)
       at android.os.Handler.dispatchMessage(Handler.java:109)
       at android.os.Looper.loop(Looper.java:207)
       at android.app.ActivityThread.main(ActivityThread.java:7539)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:958)

**Comment by christocracy** [comment]
Yes, always use the most recent version.  

The reason for your crashes are already fixed.

Read the CHANGELOG.

**Comment by UnSket** [comment]
Was it fixed in 3.0.4 version? I just want to be shure @christocracy

**Comment by christocracy** [comment]
And yes, Huawei are the worst devices on earth from an engineering perspective.

Soon these devices won’t work at all with the Background Geolocation SDK if Huawei is  denied access to Google Play Services in August.

**Comment by christocracy** [comment]
I don’t know when it was fixed, I might not have documented it in CHANGELOG, but I recall making a change [mentioned above to guard against null task](https://github.com/transistorsoft/react-native-background-geolocation/issues/699#issuecomment-502145310)

**Comment by UnSket** [comment]
Thank you! We will try to upgrade the plugin and write you if problem still there
